src/agent/src/mount.rs

```diff
@@ -573,7 +632,69 @@ async fn virtio_blk_storage_handler(
         let dev_path = get_virtio_blk_pci_device_name(sandbox, &pcipath).await?;
         storage.source = dev_path;
     }
-
+    let options = storage.options();
+    if options.contains(&"io.katacontainers.image.tarfs".to_string()) {
+        let cid = cid.ok_or_else(|| anyhow!("No container id"))?;
+        let verity_options = options[2].clone();
+        let mount_path = format!(
+            "{}/{}/{}/{}",
+            crate::rpc::CONTAINER_BASE,
+            cid,
+            "lowerdir",
+            options[4].clone()
+        );
+        let mount_type = storage.fstype();
+        if !fs::metadata(&mount_path).is_ok() {
+            fs::create_dir_all(&mount_path)
+                .map_err(anyhow::Error::from)
+                .context("Could not create mountpath")?;
+        }
+        let verity_device_name =
+            match image_rs::verity::get_verity_device_name(verity_options.as_str()) {
+                Ok(name) => name,
+                Err(err) => {
+                    return Err(anyhow!("Failed to get verity device name {}", err));
+                }
+            };
+        let sandbox = sandbox.clone();
+        let mut s = sandbox.lock().await;
+        if s.storages.contains_key(&verity_device_name) {
+            nix::mount::mount(
+                Some(format!("/dev/mapper/{}", verity_device_name).as_str()),
+                mount_path.as_str(),
+                Some(mount_type),
+                MsFlags::MS_RDONLY,
+                None::<&str>,
+            )?;
+        } else {
+            image_rs::image::mount_image_block_with_integrity(
+                verity_options.as_str(),
+                Path::new(&storage.source),
+                Path::new(&mount_path),
+                mount_type,
+            )?;
+        }
+        let entry = s
+            .container_mounts
+            .entry(cid.to_string())
+            .or_insert_with(Vec::new);
+        entry.push(mount_path.to_owned());
+        entry.push(verity_device_name.to_owned());
+        let new_storage =
+            s.set_sandbox_storage(mount_path.to_owned().as_str(), ObjectKind::MountPoint);
```

Comment:
`to_owned` allocates memory for and creates a copy of the string. You call this, then you call `as_str` to get a reference to it. It makes no sense.